### PR TITLE
fix: ignore non-route endpoints in scope auth

### PR DESCRIPTION
**Pull request by martinothamar-agent**
## Summary
- skip scope endpoint processing entirely when the app has no custom `apiScopes` configured
- ignore non-route ASP.NET endpoints when building scope metadata for apps that do use custom scopes
- add regression coverage for a plain `Microsoft.AspNetCore.Http.Endpoint` in the endpoint table

## Why
A BRG app branch upgraded to app-lib `8.12.0` failed during startup with `Unexpected endpoint type: Microsoft.AspNetCore.Http.Endpoint`. The app has no `apiScopes`, so scope authorization would no-op at request time, but initialization still scanned all endpoints and assumed every endpoint from `EndpointDataSource` was a `RouteEndpoint`. ASP.NET can expose non-route endpoints; those do not have route pattern or HTTP method metadata that this scope mapping can represent.

## Testing
- `dotnet csharpier format src/Altinn.App.Api/Infrastructure/Middleware/ScopeAuthorizationMiddleware.cs test/Altinn.App.Api.Tests/Middleware/ScopeAuthorizationServiceTests.cs`
- `dotnet test test/Altinn.App.Api.Tests/Altinn.App.Api.Tests.csproj --filter ScopeAuthorizationServiceTests --no-restore`
- `dotnet test test/Altinn.App.Api.Tests/Altinn.App.Api.Tests.csproj --no-restore`

cc @martinothamar

## Release Notes

* **Bug Fixes**
  * Enhanced scope-based authorization middleware to handle endpoint processing more robustly, reducing false authorization failures in edge cases.

* **Tests**
  * Added comprehensive test coverage for authorization service initialization across multiple endpoint scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Altinn/app-lib-dotnet/pull/1759?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by martinothamar** [comment]
what kind of endpoint type was emerging from EndpointDataSource  in the specific brg app in this case? I know I looked into types from controllers and minimal APIs, not sure what else might pop up. Also to make sure I understand, does this exception break the app or does the app still function even if initialization here fails?

**Comment by martinothamar-agent** [comment]
I reproduced this with a temporary endpoint dump registered on `ApplicationStarted` in the BRG `origin/UTVINN-2517` worktree. The endpoint type is exactly `Microsoft.AspNetCore.Http.Endpoint` from `Microsoft.AspNetCore.Routing.CompositeEndpointDataSource`.

It is not an exotic minimal API endpoint. In this upgraded branch the endpoint table contains plain base `Endpoint` entries for MVC controller actions, paired with the normal `RouteEndpoint` for the same action. The first one in the dump was:

```text
ENDPOINT type=Microsoft.AspNetCore.Http.Endpoint route=<not route endpoint> display=Altinn.App.Features.Signing.Resolution.SigneeSelectionController.UpdateSigneeAsync (Altinn.Application)
```

and there are many others, both app-defined controllers and app-lib controllers, for example:

```text
ENDPOINT type=Microsoft.AspNetCore.Http.Endpoint route=<not route endpoint> display=Altinn.App.Controllers.AdresseController.GetKommuner (Altinn.Application)
ENDPOINT type=Microsoft.AspNetCore.Http.Endpoint route=<not route endpoint> display=Altinn.App.Api.Controllers.DataController.Create (Altinn.App.Api)
ENDPOINT type=Microsoft.AspNetCore.Http.Endpoint route=<not route endpoint> display=Altinn.App.Api.Controllers.InstancesController.Get (Altinn.App.Api)
```

The metadata on these includes `ControllerActionDescriptor`, `HttpMethodMetadata`, filters/attributes, etc., but no route pattern. The corresponding `RouteEndpoint` appears separately and has `RouteEndpointBuilder+RouteDiagnosticsMetadata` and the actual route pattern. So for scope authorization, the plain base endpoints cannot be represented as our `ApiEndpoint` key anyway; only the `RouteEndpoint` carries the route we need.

On impact: it does not always immediately kill the process in local dev. In the repro the app logs `Application started`, then `ApplicationLifetime` logs a critical `AggregateException` from the `ApplicationStarted` callback. The service initialization task is then faulted. Since `ScopeAuthorizationMiddleware` calls `EnsureInitialized()` for authenticated requests, those requests would rethrow the initialization failure and fail. For this BRG app there are no `apiScopes`, so the intended behaviour is no-op; the current code still scans endpoints and leaves scope auth broken before it can no-op.

That is why I think both parts of this change matter: skip the endpoint scan entirely when no custom scopes are configured, and ignore non-route endpoints when scopes are configured because they are not mappable to route/method scope metadata.

**Comment by martinothamar** [comment]
Wut I dont understand

> The first one in the dump was:
> `ENDPOINT type=Microsoft.AspNetCore.Http.Endpoint route=<not route endpoint> display=Altinn.App.Features.Signing.Resolution.SigneeSelectionController.UpdateSigneeAsync (Altinn.Application)`

That is an endpoint that is built into app-lib, so its not app code that is provoking this condition? 
I'm confused because we do have integrations for this functionality in test/Altinn.App.Integration.Tests/CustomScopes so I would have thought that it would catch any endpoint metadata we didnt expect here and break integration tests?

**Comment by martinothamar-agent** [comment]
Yes, good catch. My previous wording was sloppy. The controller action display name is not the thing that provokes it. The trigger in the BRG app is the fallback/dynamic MVC routing path:

```csharp
app.MapControllers();
app.MapFallbackToController("Index", "Home");
```

`MapFallbackToController` sets `ControllerActionEndpointDataSource.CreateInertEndpoints = true`. ASP.NET Core documents that flag in code as creating "inert" / non-routable endpoints for dynamic selection. Then `ActionEndpointFactory` adds one `InertEndpointBuilder` endpoint per controller action, and that builder returns a plain `Microsoft.AspNetCore.Http.Endpoint`, not a `RouteEndpoint`.

Relevant ASP.NET Core source refs:
- `MapFallbackToController` sets `CreateInertEndpoints = true`: https://github.com/dotnet/aspnetcore/blob/v8.0.0/src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs#L180-L194
- `CreateInertEndpoints` comment says these are inert/non-routable endpoints for dynamic selection: https://github.com/dotnet/aspnetcore/blob/v8.0.0/src/Mvc/Mvc.Core/src/Routing/ControllerActionEndpointDataSource.cs#L43-L47
- `ActionEndpointFactory` creates `InertEndpointBuilder` when that flag is true: https://github.com/dotnet/aspnetcore/blob/v8.0.0/src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs#L59-L66
- `InertEndpointBuilder.Build()` returns base `Endpoint`: https://github.com/dotnet/aspnetcore/blob/v8.0.0/src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs#L547-L552

So the dump showing `Altinn.App.Api.Controllers.*` display names just means MVC created inert selector candidates for every controller action, including app-lib controllers and app controllers. The specific action is incidental. The actual routeable endpoint is still present separately as a `RouteEndpoint` with the route pattern.

That also explains why `test/Altinn.App.Integration.Tests/CustomScopes` did not catch this: the test app uses `UseAltinnAppCommonConfiguration()`, which maps controllers, but it does not call `MapFallbackToController` / dynamic controller routing. So `CreateInertEndpoints` stays false and the endpoint table only has the normal route endpoints.

This makes the fix less about unknown endpoint metadata from controllers/minimal APIs and more about not assuming all entries in `EndpointDataSource` are routeable endpoints. For scope auth we only have a meaningful key for `RouteEndpoint` because `ApiEndpoint` needs the route pattern.

**Comment by martinothamar** [comment]
OK, what is important in this code is that we either cover all exposed HTTP endpoints according to configuration or fail early. If the specific endpoint we found is non-routable, is it possible to detect that specifically? Is it out of the question that there may be some non-`RouteEndpoint` routable endpoint in the future that suddenly just gets skipped due to these silent skips?

I'm not convinced we should exit early based on `if (!HasDefinedCustomScopes)`, I dont see that as the root issue here

**Comment by martinothamar-agent** [comment]
Agreed. I tightened the PR to keep that invariant.

The code no longer skips arbitrary non-`RouteEndpoint` instances, and I removed the `!HasDefinedCustomScopes` early exit. We still build the endpoint metadata table on startup so unexpected endpoint shapes fail early regardless of whether custom scopes are configured.

The only non-route endpoint shape skipped now is the MVC inert endpoint produced by `MapFallbackToController`/dynamic controller routing, and only when it is paired with a normal `RouteEndpoint` for the same `ControllerActionDescriptor`. Everything else still throws `Unexpected endpoint type`.

Concretely:
- `Endpoint` + `ControllerActionDescriptor` + matching routed controller action => skip as non-routable MVC inert endpoint
- any other non-`RouteEndpoint` => fail early
- all `RouteEndpoint`s are still processed into the scope metadata lookup

I also added a regression test that proves an unexpected plain `Endpoint` still throws, plus tests for skipping only the paired MVC inert endpoint with and without custom scopes.

Local verification after the change:

```text
dotnet test test/Altinn.App.Api.Tests/Altinn.App.Api.Tests.csproj --filter ScopeAuthorizationServiceTests --no-restore
Passed: 3

dotnet test test/Altinn.App.Api.Tests/Altinn.App.Api.Tests.csproj --no-restore
Passed: 416
```

**Comment by martinothamar** [comment]
Make sure this latest iteration is tested against the app in question please

File: test/Altinn.App.Api.Tests/Middleware/ScopeAuthorizationServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            "GET"
+        );
+        var inertEndpoint = CreateMvcInertEndpoint(actionDescriptor, "GET");
+        var service = CreateService(new ApplicationMetadata("ttd/app"), inertEndpoint, routeEndpoint);
+
+        await service.EnsureInitialized();
+
+        Assert.False(service.HasDefinedCustomScopes);
+        Assert.Single(service.Metadata);
+    }
```
[the remark this answers is not in the document]
Comment (martinothamar-agent) [reply test/Altinn.App.Api.Tests/Middleware/ScopeAuthorizationServiceTests.cs:38]: Not applying this one. The scope service intentionally still scans endpoints when `ApiScopes` is absent, because the invariant here is that we either understand all exposed HTTP endpoints or fail during startup. Short-circuiting the no-scope case would have hidden the BRG failure mode again. The latest update keeps that behaviour and narrows the allowed non-`RouteEndpoint` shape to MVC inert endpoints represented by base `Endpoint` instances with `ControllerActionDescriptor` metadata.

**Comment by martinothamar-agent** [comment]
Updated and tested against the BRG app branch.

What changed after the app test:
- The first BRG validation still failed after the previous iteration on `AuthorizationController.ValidateSelectedParty`. That endpoint is also a base `Microsoft.AspNetCore.Http.Endpoint` with MVC `ControllerActionDescriptor` metadata, but it has no matching `RouteEndpoint` because it is not exposed through the attribute route table.
- I adjusted the skip to cover MVC inert endpoints as base `Endpoint` instances with `ControllerActionDescriptor` metadata, while still throwing for arbitrary non-route endpoints.
- Tests now cover an unpaired MVC inert endpoint, so this specific BRG shape is represented.

Validation:
- `dotnet test test/Altinn.App.Api.Tests/Altinn.App.Api.Tests.csproj --filter ScopeAuthorizationServiceTests --no-restore` -> passed, 3/3.
- `dotnet test test/Altinn.App.Api.Tests/Altinn.App.Api.Tests.csproj --no-restore` -> passed, 416/416.
- Repacked patched 8.12 app-lib locally and restored BRG `brg/virksomhetsregistrering` branch `UTVINN-2517` against `Altinn.App.Api/8.12.0-brgtest.11` and `Altinn.App.Core/8.12.0-brgtest.11`.
- Started the BRG app with `ASPNETCORE_ENVIRONMENT=Development dotnet run --project App/App.csproj --no-launch-profile --no-restore`.
- Scope auth initialized successfully: `Scope authorization initialized. Processed 118 endpoints`.

The remaining runtime warnings/errors in that run are localtest connection refused on `localhost:5101`, after the app had already started and scope auth had initialized.

**Comment by martinothamar** [comment]
> The first BRG validation still failed after the previous iteration on AuthorizationController.ValidateSelectedParty

Now youve lost me again, this is our endpoint right? What makes it troublesome in the brg-app but not in our own tests. Or did I lose the plot inbetween the force pushes here

**Comment by martinothamar-agent** [comment]
Yes, it is our endpoint. The BRG-specific part is not the controller action itself, it is how the app wires endpoint routing.

BRG has `app.MapFallbackToController("Index", "Home")`. That makes MVC add a second set of non-routable/inert MVC endpoint candidates to `EndpointDataSource` for fallback/dynamic controller selection. Those inert candidates are plain `Microsoft.AspNetCore.Http.Endpoint`, not `RouteEndpoint`, but they still carry `ControllerActionDescriptor` metadata and display names like real controller actions.

My previous iteration only skipped inert endpoints when I could prove they were paired with a routable `RouteEndpoint` for the same action descriptor. The BRG run then found `Altinn.App.Api.Controllers.AuthorizationController.ValidateSelectedParty` as one of those inert base `Endpoint`s, but not in a shape that matched that pairing check. So the troublesome bit is not that `ValidateSelectedParty` is exposed differently in BRG; it is that BRG's fallback route causes MVC to put this extra inert endpoint object into the endpoint data source, and the scope scanner was treating any non-`RouteEndpoint` as fatal.

Our existing tests missed it because they only covered controller/minimal route endpoints, and my first new unit test still modelled the inert endpoint as paired with a route endpoint. The latest test now models the unpaired inert MVC shape directly: base `Endpoint` + `ControllerActionDescriptor` metadata. Arbitrary base endpoints without MVC action metadata still throw, so the fail-early invariant is preserved for unknown endpoint shapes.

**Review by martinothamar — APPROVED** [review]
